server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java

```diff
@@ -0,0 +1,714 @@
+
+/**
+ * Base collector manager that creates and reduces {@link QueryPhaseCollector}s used in the query phase.
+ * It is aware of profiling, in that it applies the proper wrapping as well as reduction of collectors when profiling is enabled.
+ * The different subclasses plug in their specific behaviour that revolves around top docs collection.
+ */
+abstract class QueryPhaseCollectorManager implements CollectorManager<Collector, QueryPhaseResult> {
+    private final Weight postFilterWeight;
+    private final QueryPhaseCollector.TerminateAfterChecker terminateAfterChecker;
+    private final CollectorManager<? extends Collector, Void> aggsCollectorManager;
+    private final Float minScore;
+    private final boolean profile;
+
+    QueryPhaseCollectorManager(
+        Weight postFilterWeight,
+        QueryPhaseCollector.TerminateAfterChecker terminateAfterChecker,
+        CollectorManager<? extends Collector, Void> aggsCollectorManager,
+        Float minScore,
+        boolean profile
+    ) {
+        this.postFilterWeight = postFilterWeight;
+        this.terminateAfterChecker = terminateAfterChecker;
+        this.aggsCollectorManager = aggsCollectorManager;
+        this.minScore = minScore;
+        this.profile = profile;
+    }
+
+    @Override
+    public final Collector newCollector() throws IOException {
+        Collector aggsCollector = null;
+        if (aggsCollectorManager != null) {
+            aggsCollector = aggsCollectorManager.newCollector();
+            if (profile) {
+                aggsCollector = new InternalProfileCollector(aggsCollector, REASON_AGGREGATION);
+            }
+        }
+
+        Collector topDocsCollector = newTopDocsCollector();
+        if (profile) {
+            topDocsCollector = new InternalProfileCollector(topDocsCollector, getTopDocsProfilerName());
+        }
+
+        QueryPhaseCollector queryPhaseCollector = new QueryPhaseCollector(
+            topDocsCollector,
+            postFilterWeight,
+            terminateAfterChecker,
+            aggsCollector,
+            minScore
+        );
+        if (profile) {
+            if (aggsCollector == null) {
+                return new InternalProfileCollector(
+                    queryPhaseCollector,
+                    REASON_SEARCH_QUERY_PHASE,
+                    (InternalProfileCollector) topDocsCollector
+                );
+            }
+            return new InternalProfileCollector(
+                queryPhaseCollector,
+                REASON_SEARCH_QUERY_PHASE,
+                (InternalProfileCollector) topDocsCollector,
+                (InternalProfileCollector) aggsCollector
+            );
+        }
+        return queryPhaseCollector;
+    }
+
+    protected abstract Collector newTopDocsCollector() throws IOException;
+
+    @Override
+    public final QueryPhaseResult reduce(Collection<Collector> collectors) throws IOException {
```

Comment:
The casting required in this method is perhaps the only cons I see around removing collector managers wrapping, as you can have many different types of collectors created, that need to be reduced. On the other hand, the structure is quite predictable, and it depends on the logic that newCollector includes above. I think that is reasonable and the pros of removing the wrapping outweigh the need for casting collectors in this method.

Reply:
It's not ideal, but I agree with the trade-off here to significantly reduce the abstractions.

Reply:
This also comes from the collector manager design: you create collector, and get the same ones as arguments to reduce. It would probably make things easier to just keep track of the collectors that were created and just reduce those, then you don't need casting etc. that feels though not very aligned with the Lucene design. Perhaps this is something that we could look at simplifying in Lucene.